### DISCUSSION TITLE
Custom kernel versions

### DISCUSSION THREAD

**Discussion by PowerUser64** [nix-community/raspberry-pi-nix#44] [Q&A] [answered]
How are we supposed to set custom kernel versions when using rpi-nix? I would like to run a newer version that has in-tree support for bcachefs (kernel 6.7+), but I realize that this project sets the kernel version for me. Is there a way I can get a newer kernel version without breaking my system?

I guess another question would be if it's possible to use the official (non-rpi foundation) linux kernels on my pi with or without this project.

Edit: I found an option called [`raspberry-pi-nix.kernel-version`](https://github.com/nix-community/raspberry-pi-nix/blob/6d61a9b18b1d3786ebd1cc73b7a3b37959bf4557/rpi/default.nix#L15) that isn't mentioned in the readme. I set it to `v6_10_0-rc5`, but it seems like the raspberry pi foundation has a branch for 6.11 available, which I would like to use even if I have to compile. Maybe I can override the [input](https://github.com/nix-community/raspberry-pi-nix/blob/6d61a9b18b1d3786ebd1cc73b7a3b37959bf4557/flake.nix#L14-L17) for this kernel version to make it point to 6.11?

**Comment by PowerUser64** [comment]
@tstat would you know the answer to this?

**Reply by tstat** [reply, marked as answer]
Yes, I expect that your edit would work. You also don't need to override the input, but can instead add to the `pkgs.rpi-kernels` dictionary with your own `version/board` pair and set those options in `raspberry-pi-nix.kernel-version` and `raspberry-pi-nix.board` options to match.

This project is arranged to support multiple kernel/board versions ([see here](https://github.com/nix-community/raspberry-pi-nix/blob/master/overlays/default.nix#L11-L45)), so a PR to add 6.11 as a kernel option that tracks the rpi linux branch would also be accepted.

> I guess another question would be if it's possible to use the official (non-rpi foundation) linux kernels on my pi with or without this project.

This is possible, with various levels of success. You will find plenty of threads about it on the nixpkgs issue tracker. This project is focused on stability and full hardware support, and hence uses the raspberry pi linux fork. So, this repo without the rpi linux fork doesn't make much sense.

**Reply by PowerUser64** [reply]
Cool, thanks for the info!

**Reply by PowerUser64** [reply]
Oh, and two other things,

> >  [is it] possible to use the official (non-rpi foundation) linux kernels?
>
> This is possible, with various levels of success.

If I wanted to do this, would I just set kernelPackages like how I do it on other systems?

> You will find plenty of threads about it on the nixpkgs issue tracker.

I see. I guess you mean issues like https://github.com/NixOS/nixpkgs/issues/300328, https://github.com/NixOS/nixpkgs/issues/142848, https://github.com/NixOS/nixpkgs/issues/82455, and https://github.com/NixOS/nixpkgs/issues/97064? (Web search for "nixpkgs raspberry pi kernel issue".) A lot of mentions of bit failures. I guess that does sound pretty sketchy. I was hoping it would just be stuff like graphics bugs since my rpi is headless 🫠
